src/util/match.h

```diff
@@ -7,33 +7,36 @@
 
 #include <variant>
 
-// Helper for using `std::visit` with Rust-style match syntax.
-//
-// This can be used in place of defining an explicit visitor. `std::visit` requires an
-// exhaustive match; to emulate Rust's catch-all binding, use an `(auto arg)` template
-// operator().
-//
-// Care must be taken that implicit conversions are handled correctly. For instance, a
-// `(double arg)` operator() *will also* bind to `int` and `long`, if there isn't an
-// earlier satisfying match.
-//
-// This corresponds to visitor example #4 in the `std::visit` documentation:
-// https://en.cppreference.com/w/cpp/utility/variant/visit
+/// Helper for using `std::visit` with Rust-style match syntax.
+///
+/// This can be used in place of defining an explicit visitor. `std::visit` requires an
+/// exhaustive match; to emulate Rust's catch-all binding, use an `(auto arg)` template
+/// operator().
+///
+/// Care must be taken that implicit conversions are handled correctly. For instance, a
+/// `(double arg)` operator() *will also* bind to `int` and `long`, if there isn't an
+/// earlier satisfying match.
+///
+/// This corresponds to visitor example #4 in the `std::visit` documentation:
+/// https://en.cppreference.com/w/cpp/utility/variant/visit#Example, where it is named `overloaded`.
 template<class... Ts> struct match : Ts... { using Ts::operator()...; };
-// explicit deduction guide (not needed as of C++20)
+
+/// explicit deduction guide (not needed as of C++20)
 template<class... Ts> match(Ts...) -> match<Ts...>;
 
-// A wrapper around two-argument `std::visit` that reverses the arguments, putting the
-// value to be visited first. This is normally used as:
-//
-// examine(specimen, match {
-//     ...
-// })
-//
-// The return type is inferred as it would be for `std::visit`.
-template<class Specimen, class Visitor>
-decltype(auto) examine(Specimen&& specimen, Visitor&& visitor) {
-    return std::visit(visitor, specimen);
+/// A wrapper around two-argument `std::visit` that reverses the arguments, putting the
+/// value to be visited first. This is normally used as:
+///
+///     examine(specimen,
+///         [](const Foo& foo) { … },
+///         [](const Bar& bar) { … },
+///         ...
+///     );
+///
+/// The return type is inferred as it would be for `std::visit`.
+template<class Specimen, class... Visitors>
+constexpr decltype(auto) examine(Specimen&& specimen, Visitors&&... visitors) {
```

Comment:
```suggestion
decltype(auto) examine(Specimen&& specimen, Visitors&&... visitors) {
```
This function doesn't meet the [requirements for `constexpr` functions](https://en.cppreference.com/w/cpp/language/constexpr#constexpr_function), specifically that "its return value (if any) and each of its parameters must be of a [LiteralType](https://en.cppreference.com/w/cpp/named_req/LiteralType)".

(`std::visit` also doesn't meet that requirement and is `constexpr`, but maybe it's magic.)

Reply:
Oh, I missed this part:
> For `constexpr` function templates and `constexpr` member functions of class templates, at least one specialization must satisfy the abovementioned requirements. Other specializations are still considered as `constexpr`, even though a call to such a function cannot appear in a constant expression. If no specialization of the template would satisfy the requirements for a `constexpr` function when considered as a non-template function, the template is ill-formed, no diagnostic required. (until C++23)

It is unclear whether the last sentence means "If no possible specialization" (which would be okay as-is), or "If no actually instantiated specialization" &mdash; in which case we'd have to include an example like this:

```c++
static inline void dummy() {
    static_assert(42 == examine(42, [](int x) constexpr { return x; }));
}
```
It makes very little sense that you would need to do that, but this is C++, ¯\_(ツ)_/¯.

Reply:
@str4d, @nuttycom and I talked this through and convinced ourselves that:
* the intended interpretation is probably "If no possible specialization";
* but, it is a good idea to include the above dummy function anyway, since it acts as a static test that the `constexpr` usage works.

Reply:
@str4d also pointed out that `match` would be a better name for `examine` as of this PR. `match` would then be renamed to something else, say `matcher`. (Outside this file, currently there is one nontrivial "bare" use of `match` in `wallet/rpcwallet.cpp`, and one other that is trivially replaceable with `examine` in `wallet/wallet_tx_builder.cpp`.)

Reply:
`overloaded` is the name used in [the example we copied it from](https://en.cppreference.com/w/cpp/utility/variant/visit#Example). I don’t think it’s a good name, but I think removing the only remaining explicit `match` from rpcwallet.cpp should be easy, so then the name is pretty internal.

Reply:
If you look at the example in `rpcwallet.cpp`:
https://github.com/zcash/zcash/blob/b072a8950551967e40f5f3360359813b3b3b2150/src/wallet/rpcwallet.cpp#L585-L595
the name `matcher` makes a lot of sense (more than `overloaded`). It's basically just treating the set of match clauses as a first-class object. And I don't think we need to discourage this construct from being used directly; it's sometimes the more natural approach to avoid code duplication.